---
title: "Evaluate performance of whitening transform"
subtitle: 'Run on gene expression from GTEx'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: true
    smart: true
params:
  DATASET: NULL
---


<!--- 

cd /hpc/users/hoffmg01/work/decorrelate_analysis

git pull 
R

system("git pull"); rmarkdown::render("gtex.Rmd")



# https://hoffmg01.hpc.mssm.edu/decorrelate_analysis


--->


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = TRUE)
```


```{r load.packages, cache=FALSE}
suppressPackageStartupMessages({
library(ggplot2)
library(tidyverse)
library(decorrelate)
library(Rfast)
library(corpcor)
library(ShrinkCovMat)
library(jsonlite)
library(cowplot)
library(CovTools)
library(RhpcBLASctl)
})
omp_set_num_threads(6)
```

# Load data
```{r load, cache=FALSE, eval=FALSE}
file = "GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz"
dfTMP <- read.delim(file, as.is=TRUE, row.names=1, check.names=FALSE ,skip=2)
gene.names.df <- dfTMP[, 'Description', drop=FALSE]
dfTMP <- dfTMP[, !(names(dfTMP) %in% c('Description'))]

file = "GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"
info <- read.delim(file, as.is=TRUE, header=TRUE, row.names=1)

save(dfTMP, info, file="GTEx_data.RDATA")
```

```{r functions, cache=FALSE}
if( ! "dfTMP" %in% ls()){
  load("GTEx_data.RDATA")
}

df_colors = fromJSON("colors.json")
df_colors = lapply(df_colors, data.frame)
df_colors = do.call(rbind, df_colors)

# norm of matrix compared to identity,
# This is the root mean squared error of the off diagonal entries
normCov = function(Sigma){

  if( length(Sigma) == 1)
    if( is.na(Sigma) ) return(NA)
  # base::norm(Sigma - diag(1, nrow(Sigma)), "F")^2 / length(Sigma)

  mse = mean((Sigma-diag(1, nrow(Sigma)))^2)
  sqrt(mse)
}

# matrix square root
msqrt = function(S){
  dcmp = eigen(S)
  # with(dcmp, vectors %*% diag(sqrt(values)) %*% t(vectors))
  with(dcmp, vectors %*% (sqrt(values) * t(vectors)))
}

minvsqrt = function(S, tol = 1e-14){
  dcmp = eigen(S)  
  k = sum(dcmp$values > tol)
  # with(dcmp, vectors %*% diag(1/sqrt(values)) %*% t(vectors))
  with(dcmp, vectors[,seq(k)] %*% ((1/sqrt(values[seq(k)])) * t(vectors[,seq(k)])))
}


# whiten with pseudoinverse with fixed rank
get_w_ginv = function(X, k, tol = 1e-14){
  C = cov(X)
  dcmp = eigen(C)

  k = sum(dcmp$values[seq(k)] > tol)
  # W = with(dcmp, vectors[,seq(k)] %*% diag(1/sqrt(values[seq(k)])) %*% t(vectors[,seq(k)]))
  W = with(dcmp, vectors[,seq(k)] %*% ((1/sqrt(values[seq(k)])) * t(vectors[,seq(k)])))
  W
}
```



```{r analysis}
tab = sort(table(info$SMTSD))

# for each tissue
res = lapply(tab[1:13], function(tissue){

  cat("\r", tissue, "        ")

  # subset info
  infoSub = info[info$SMTSD == tissue,]
  include = rownames(infoSub)[rownames(infoSub) %in% colnames(dfTMP)]

  # keep tissues with at least 200 samples
  if( length(include) < 200) return(NULL)

  # subset gene expression, convert to log2 TPM
  geneExpr = dfTMP[,include] + 1e-3

  # keep genes with at least 1 TMP in 50 samples
  keep = rowSums(geneExpr > 1) > 50

  # get log2 TMP with features on columns
  Ys = scale(log2(t(geneExpr[keep,])))

  n = nrow(Ys)

  idx_train = sort(sample.int(n, n/2, replace=FALSE))
  df = data.frame()

  # Eval methods
  #-------------

  # decorrelate
  tm = system.time({
    ecl <- eclairs( Ys[idx_train,])
    y_white <- decorrelate(Ys[-idx_train,], ecl)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = 'GIW-EB', 
                      t(c(tm)),
                      rmse = rmse))

  # decorrelate
  tm = system.time({
    ecl <- eclairs( Ys[idx_train,], k=50)
    y_white <- decorrelate(Ys[-idx_train,], ecl)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = 'GIW-EB (k=50)', 
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    ecl <- eclairs( Ys[idx_train,])
    y_white <- decorrelate(Ys[-idx_train,], ecl, lambda = 0)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "lambda = 0",  
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    ecl <- eclairs( Ys[idx_train,])
    y_white <- decorrelate(Ys[-idx_train,], ecl, lambda = 0.01)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "lambda = 0.01", 
                      t(c(tm)),
                      rmse = rmse))

  # tm = system.time({
  #   ecl <- eclairs( Ys[idx_train,])
  #   y_white <- decorrelate(Ys[-idx_train,], ecl, lambda = 1e-4)
  #   })   
  # rmse = normCov(cora(y_white))
  # df = rbind(df, data.frame(
  #                     Method = "lambda = 1e-4",  
  #                     t(c(tm)),
  #                     rmse = rmse))

  # tm = system.time({
  #   fit <- CovTools::CovEst.2003LW( Ys[idx_train,] )
  #   y_white <- Ys[-idx_train,] %*% minvsqrt(fit$S)
  #   })   
  # rmse = normCov(cora(y_white))
  # df = rbind(df, data.frame(
  #                     Method = "Ledoit-Wolf",  
  #                     t(c(tm)),
  #                     rmse = rmse))

  # tm = system.time({
  #   fit <- CovTools::CovEst.2010OAS( Ys[idx_train,] )
  #   y_white <- Ys[-idx_train,] %*% minvsqrt(fit$S)
  #   })   
  # rmse = normCov(cora(y_white))
  # df = rbind(df, data.frame(
  #                     Method = "OAS",  
  #                     t(c(tm)),
  #                     rmse = rmse))

  tm = system.time({
    fit <- shrinkcovmat.equal( t(Ys[idx_train,]) )
    y_white <- Ys[-idx_train,] %*% minvsqrt(fit$Sigmahat)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "Touloumis", 
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    C <- cor.shrink( Ys[idx_train,], verbose=FALSE )
    y_white <- Ys[-idx_train,] %*% minvsqrt(C)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "SchÃ¤fer-Strimmer", 
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    # learn transformation
    k <- min(dim(Ys[idx_train,]))-1
    W <- get_w_ginv(Ys[idx_train,], k)
    y_white <- tcrossprod(Ys[-idx_train,], W)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "Pseudoinverse", 
                      t(c(tm)),
                      rmse = rmse))    

  df$Tissue = tissue
  df$n = nrow(Ys)
  df$p = ncol(Ys)
  df$rMSE_baseline = normCov(cora(Ys[-idx_train,]))
  df
})
res = do.call(rbind, res)
```


















