---
title: "Evaluate performance of whitening transform"
subtitle: 'Run on gene expression from GTEx'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: true
    smart: true
params:
  DATASET: NULL
---


<!--- 

cd /hpc/users/hoffmg01/work/decorrelate_analysis

git pull 
R

system("git pull"); rmarkdown::render("gtex.Rmd")



# https://hoffmg01.hpc.mssm.edu/decorrelate_analysis


--->


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = TRUE)
```


```{r load.packages, cache=FALSE}
suppressPackageStartupMessages({
library(ggplot2)
library(tidyverse)
library(decorrelate)
library(Rfast)
library(corpcor)
library(parallel)
library(ShrinkCovMat)
library(jsonlite)
library(cowplot)
library(CovTools)
library(RhpcBLASctl)
})
omp_set_num_threads(6)
```

# Load data
```{r load, cache=FALSE, eval=FALSE}
file = "GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz"
dfTMP <- read.delim(file, as.is=TRUE, row.names=1, check.names=FALSE ,skip=2)
gene.names.df <- dfTMP[, 'Description', drop=FALSE]
dfTMP <- dfTMP[, !(names(dfTMP) %in% c('Description'))]

file = "GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"
info <- read.delim(file, as.is=TRUE, header=TRUE, row.names=1)

save(dfTMP, info, file="GTEx_data.RDATA")
```

```{r functions, cache=FALSE}
if( ! "dfTMP" %in% ls()){
  load("GTEx_data.RDATA")
}

df_colors = fromJSON("colors.json")
df_colors = lapply(df_colors, data.frame)
df_colors = do.call(rbind, df_colors)

# norm of matrix compared to identity,
# This is the root mean squared error of the off diagonal entries
normCov = function(Sigma){

  if( length(Sigma) == 1)
    if( is.na(Sigma) ) return(NA)
  # base::norm(Sigma - diag(1, nrow(Sigma)), "F")^2 / length(Sigma)

  mse = mean((Sigma-diag(1, nrow(Sigma)))^2)
  sqrt(mse)
}

# matrix square root
msqrt = function(S){
  dcmp = eigen(S)
  # with(dcmp, vectors %*% diag(sqrt(values)) %*% t(vectors))
  with(dcmp, vectors %*% (sqrt(values) * t(vectors)))
}

minvsqrt = function(S, tol = 1e-14){
  dcmp = eigen(S)  
  k = sum(dcmp$values > tol)
  # with(dcmp, vectors %*% diag(1/sqrt(values)) %*% t(vectors))
  with(dcmp, vectors[,seq(k)] %*% ((1/sqrt(values[seq(k)])) * t(vectors[,seq(k)])))
}


# whiten with pseudoinverse with fixed rank
get_w_ginv = function(X, k, tol = 1e-14){
  C = cov(X)
  dcmp = eigen(C)

  k = sum(dcmp$values[seq(k)] > tol)
  # W = with(dcmp, vectors[,seq(k)] %*% diag(1/sqrt(values[seq(k)])) %*% t(vectors[,seq(k)]))
  W = with(dcmp, vectors[,seq(k)] %*% ((1/sqrt(values[seq(k)])) * t(vectors[,seq(k)])))
  W
}

colMethods = c("GIW-EB" = "red",
              "GIW-EB (k=50)" = "#9f1214",
              "lambda = 0.01" = "#377EB8", 
              "lambda = 1e-4" = "#37b8b2",
              "lambda = 0" = "#0000cd", 
              "Ledoit-Wolf" = "#FF7F00",    
              "OAS"= "#cccc29", 
              "Touloumis" = "#A65628", 
              "Schäfer-Strimmer" = "#f569b3", 
              "Pseudoinverse" = "green3",                           
              "Baseline" = "grey50",
              "Oracle" = "black")
```



```{r analysis, message=FALSE}
tab = sort(table(info$SMTSD))

# for each tissue
file.remove("progress.log")
res = mclapply( names(tab), function(tissue){

  write(tissue, file="progress.log", append=TRUE)

  # subset info
  infoSub = info[info$SMTSD == tissue,]
  include = rownames(infoSub)[rownames(infoSub) %in% colnames(dfTMP)]

  # keep tissues with at least 200 samples
  if( length(include) < 200) return(NULL)

  # subset gene expression, convert to log2 TPM
  geneExpr = dfTMP[,include] + 1e-4

  # keep genes with at least 1 TMP in 50 samples
  keep = rowSums(geneExpr > 1) > 50

  # get log2 TMP with features on columns
  Ys = scale(log2(t(geneExpr[keep,])))

  n = nrow(Ys)

  idx_train = sort(sample.int(n, n * .8, replace=FALSE))
  df = data.frame()

  # Eval methods
  #-------------

  # decorrelate
  tm = system.time({
    ecl <- eclairs( Ys[idx_train,])
    y_white <- decorrelate(Ys[-idx_train,], ecl)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = 'GIW-EB', 
                      t(c(tm)),
                      rmse = rmse))

  # decorrelate
  tm = system.time({
    ecl <- eclairs( Ys[idx_train,], k=50)
    y_white <- decorrelate(Ys[-idx_train,], ecl)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = 'GIW-EB (k=50)', 
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    ecl <- eclairs( Ys[idx_train,])
    y_white <- decorrelate(Ys[-idx_train,], ecl, lambda = 0)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "lambda = 0",  
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    ecl <- eclairs( Ys[idx_train,])
    y_white <- decorrelate(Ys[-idx_train,], ecl, lambda = 0.01)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "lambda = 0.01", 
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    ecl <- eclairs( Ys[idx_train,])
    y_white <- decorrelate(Ys[-idx_train,], ecl, lambda = 1e-4)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "lambda = 1e-4",  
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    fit <- CovTools::CovEst.2003LW( Ys[idx_train,] )
    y_white <- Ys[-idx_train,] %*% minvsqrt(fit$S)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "Ledoit-Wolf",  
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    fit <- CovTools::CovEst.2010OAS( Ys[idx_train,] )
    y_white <- Ys[-idx_train,] %*% minvsqrt(fit$S)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "OAS",  
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    fit <- shrinkcovmat.equal( t(Ys[idx_train,]) )
    y_white <- Ys[-idx_train,] %*% minvsqrt(fit$Sigmahat)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "Touloumis", 
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    C <- cor.shrink( Ys[idx_train,], verbose=FALSE )
    y_white <- Ys[-idx_train,] %*% minvsqrt(C)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "Schäfer-Strimmer", 
                      t(c(tm)),
                      rmse = rmse))

  tm = system.time({
    # learn transformation
    k <- min(dim(Ys[idx_train,]))-1
    W <- get_w_ginv(Ys[idx_train,], k)
    y_white <- tcrossprod(Ys[-idx_train,], W)
    })   
  rmse = normCov(cora(y_white))
  df = rbind(df, data.frame(
                      Method = "Pseudoinverse", 
                      t(c(tm)),
                      rmse = rmse))    

  df$Tissue = tissue
  df$n = nrow(Ys)
  df$p = ncol(Ys)
  df$rMSE_baseline = normCov(cora(Ys[-idx_train,]))
  df
}, mc.cores=12)
res = do.call(rbind, res)
res$Method = factor(res$Method, names(colMethods))
res$Method = droplevels(res$Method)
```

# Relative rMSE
```{r plots, cache=FALSE, fig.width=12, fig.height=20}
res = res %>%
  mutate(value = 100*rmse / rMSE_baseline) %>%
  mutate( Tissue = gsub("^(.*) \\(.*$", "\\1", Tissue))

res_all = res %>%
  group_by(Tissue, n, p) %>%
  reframe(rMSE_baseline = rMSE_baseline) %>%
  distinct %>%
  with(.,tibble(Method = "baseline", rmse=rMSE_baseline, Tissue, n, p)) %>%
  bind_rows(res %>% select(Method, rmse, Tissue, n,p),.)


cutoff = 59

res %>%
  ggplot(aes(Method, value, fill=Method)) + 
    geom_bar(stat="identity") +
    theme_classic()  +
    theme(aspect.ratio=1, 
      plot.title = element_text(hjust = 0.5),
      legend.position="none") +
    facet_wrap(~ Tissue, ncol=5) +
    ylab("rMSE compared to baseline (%)")  +
    coord_flip(expand=FALSE, ylim=c(0, cutoff) ) + 
    scale_fill_manual(name="Method", values = colMethods) +    
    geom_text(data=filter(res, value > cutoff), aes(Method, cutoff*.9, label=round(value, digits=1)), color="white")
```

# rMSE
```{r rMSE, cache=FALSE, fig.width=12, fig.height=20}
cutoff = 0.24

res_sub = res_all %>%
  filter(Method != "lambda = 0") %>%
  droplevels 

res_sub%>%
  ggplot(aes(Method, rmse, fill=Method)) + 
    geom_bar(stat="identity") +
    theme_classic()  +
    theme(aspect.ratio=1, 
      plot.title = element_text(hjust = 0.5),
      legend.position="none") +
    ylab("Root mean squared error")  +
    coord_flip(expand=FALSE, ylim=c(0, cutoff) ) + 
    scale_fill_manual(name="Method", values = colMethods) +
    facet_wrap(~Tissue, ncol=5) +    
    geom_text(data=filter(res_sub, rmse > cutoff), aes(Method, cutoff*.9, label=round(rmse, digits=2)), color="white")
```




```{r plots.focus, fig.width=9, fig.height=4, cache=FALSE}
cutoff = 59
res2 = res %>%
  filter(Tissue == "Brain - Hypothalamus") 

fig1 = res2 %>%
  ggplot(aes(Method, value, fill=Method)) + 
    geom_bar(stat="identity") +
    theme_classic()  +
    theme(aspect.ratio=1, 
      plot.title = element_text(hjust = 0.5),
      legend.position="none") +
    ylab("rMSE compared to baseline (%)")  +
    coord_flip(expand=FALSE, ylim=c(0, cutoff) ) + 
    scale_fill_manual(name="Method", values = colMethods) +    
    geom_text(data=filter(res2, value > cutoff), aes(Method, cutoff*.9, label=round(value, digits=1)), color="white")
fig1

cutoff = 0.3
res_sub = res_all %>%
  filter(Tissue == "Brain - Hypothalamus")  %>%
  filter(Method != "lambda = 0")

fig1a = res_sub %>%
  ggplot(aes(Method, rmse, fill=Method)) + 
    geom_bar(stat="identity") +
    theme_classic()  +
    theme(aspect.ratio=1, 
      plot.title = element_text(hjust = 0.5),
      legend.position="none") +
    ylab("Root mean squared error")  +
    coord_flip(expand=FALSE, ylim=c(0, cutoff) ) + 
    scale_fill_manual(name="Method", values = colMethods) +    
    geom_text(data=filter(res_sub, rmse > cutoff), aes(Method, cutoff*.9, label=round(rmse, digits=2)), color="white") 
fig1a
```



# Wall time
```{r time.wall, fig.width=9, fig.height=4, cache=FALSE}
fig2 = res %>% 
  group_by(Method) %>%
  summarize(elapsed = sum(elapsed) / 3600) %>%
  ggplot(aes(Method, 1000*elapsed, fill=Method)) +
    geom_bar(stat="identity") + 
    theme_bw()  +
    theme(aspect.ratio=1, 
      plot.title = element_text(hjust = 0.5), 
      legend.position = "none",
      panel.grid.major.y = element_blank(),
      panel.grid.minor.x = element_blank()) +
    scale_fill_manual(name="Method",values = colMethods) +
    coord_flip() +
    scale_y_log10(labels = ~.x/1000, expand=c(0,0), limits=c(1, 80000), breaks  =c(10, 100, 1000, 10000, 30000)) +
    ylab("Wall time (hours)")
fig2
```


```{r combine, fig.width=9, fig.height=4, cache=FALSE}
plot_grid(fig1, fig2)
plot_grid(fig1a, fig2)
```




# CPU time
```{r time.cpu, fig.width=9, fig.height=4, cache=FALSE}
res %>% 
  group_by(Method) %>%
  summarize(user.self = sum(user.self) / 3600) %>%
  ggplot(aes(Method, 1000*user.self, fill=Method)) +
    geom_bar(stat="identity") + 
    theme_bw()  +
    theme(aspect.ratio=1, 
      plot.title = element_text(hjust = 0.5), 
      legend.position = "none",
      panel.grid.major.y = element_blank(),
      panel.grid.minor.x = element_blank()) +
    scale_fill_manual(name="Method",values = colMethods) +
    coord_flip() +
    scale_y_log10(labels = ~.x/1000, expand=c(0,0), limits=c(1, 80000), breaks  =c(10, 100, 1000, 10000, 35000)) +
    ylab("CPU time (hours)")
```



















